### PR TITLE
Fix "serve" for relative public URL (Eg: ".", "./dist", "dist")

### PR DESCRIPTION
**Pull request by trongnd**

### CODE REVIEW

File: src/Server.js [hunk 1]
```diff
@@ -9,6 +9,7 @@ const prettyError = require('./utils/prettyError');
 const AnsiToHtml = require('ansi-to-html');
 const logger = require('./Logger');
 const path = require('path');
+const fs = require('fs');
```
Comment (DeMoorJasper) [inline src/Server.js]: Probably better to use the fs-utils that are part of parcel `./utils/fs`

**Comment by DeMoorJasper** [comment]
Also please add an explanation in the PR with related issues, what it solves.
Possibly also how you solved it.

**Comment by trongnd** [comment]
Yay. Sorry for missing details.
I'm trying to use relative public URL with parcel. The command would similar to:
`parcel serve src/index.html --public-url <relative-public-path>`
`<relative-public-path>` could be, e.g: `dist`, `./dist`, `.`
The built output is fine, but `serve` feature is not working correctly.

The currently logic of `serve` is:
https://github.com/parcel-bundler/parcel/pull/1997/files#diff-6f508a0628bff9c4c0d4a7bd1d5f40a5L55
```javascript
} else if (
  !pathname.startsWith(bundler.options.publicURL) ||
  path.extname(pathname) === ''
) {
   return sendIndex();
}
```
While `pathname` always starts with `/`, and `publicURL` doesn't start with `/`, so the condition is always `false`, that causes `serve` always sends `index` even on other file requests (`.js`, `.css`, etc...)
I updated so in case request URL has file extension, it would check for file existence, if the file exists, it send the file content otherwise send the index HTML.

Another issue I see that `sourceMappingURL` is using absolute path. That doesn't work with relative public URL in some cases. So I also updated to use only source map file name for `sourceMappingURL` in case public URL is relative.

File: src/Server.js [hunk 2]
```diff
@@ -51,18 +52,27 @@ function middleware(bundler) {
...
-        req.url = pathname.slice(bundler.options.publicURL.length);
-        return serve(req, res, send404);
+        // convert related publicURL path to absolute path.
+        // eg: dist -> /dist, ./dist -> /dist, /dist -> /dist
+        let publicURL = `/${bundler.options.publicURL.replace(/^\.?\/?/, '')}`;
+        if (publicURL !== '/' && pathname.startsWith(publicURL)) {
+          pathname = pathname.slice(publicURL.length);
+        }
+        // check if file exists
+        fs.exists(path.join(bundler.options.outDir, pathname.slice(1))).then(
```
Comment (DeMoorJasper) [inline src/Server.js]: Use await, it's cleaner

File: src/packagers/JSConcatPackager.js [hunk 3]
```diff
@@ -530,10 +530,10 @@ class JSConcatPackager extends Packager {
       // Add source map url if a map bundle exists
       let mapBundle = this.bundle.siblingBundlesMap.get('map');
       if (mapBundle) {
-        output += `\n//# sourceMappingURL=${urlJoin(
-          this.options.publicURL,
-          path.basename(mapBundle.name)
-        )}`;
+        let sourceMappingURL = this.options.publicURL.startsWith('/')
```
Comment (DeMoorJasper) [inline src/packagers/JSConcatPackager.js]: Why did you change this?

**Comment by devongovett** [comment]
Due to move to monorepo which rewrote the git history, this PR was automatically closed. Please open a new PR against master with your changes. You can do a diff against the `1.x` branch to see them. Sorry about that!
